exp/topology/desiredstate/desired_state.go

```diff
@@ -30,24 +31,58 @@ import (
 
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 	"sigs.k8s.io/cluster-api/controllers/external"
+	"sigs.k8s.io/cluster-api/controllers/remote"
 	expv1 "sigs.k8s.io/cluster-api/exp/api/v1beta1"
 	runtimecatalog "sigs.k8s.io/cluster-api/exp/runtime/catalog"
 	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	"sigs.k8s.io/cluster-api/exp/topology/scope"
 	"sigs.k8s.io/cluster-api/feature"
 	"sigs.k8s.io/cluster-api/internal/contract"
-	"sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/scope"
+	"sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/patches"
 	"sigs.k8s.io/cluster-api/internal/hooks"
 	tlog "sigs.k8s.io/cluster-api/internal/log"
-	"sigs.k8s.io/cluster-api/internal/topology/names"
+	runtimeclient "sigs.k8s.io/cluster-api/internal/runtime/client"
+	"sigs.k8s.io/cluster-api/internal/topology/clustershim"
+	topologynames "sigs.k8s.io/cluster-api/internal/topology/names"
+	"sigs.k8s.io/cluster-api/internal/topology/ownerrefs"
+	"sigs.k8s.io/cluster-api/internal/topology/selectors"
 	"sigs.k8s.io/cluster-api/internal/webhooks"
 	"sigs.k8s.io/cluster-api/util"
 )
 
-// computeDesiredState computes the desired state of the cluster topology.
+// Generator is a generator to generate the desired state.
+type Generator interface {
+	Generate(ctx context.Context, s *scope.Scope) (*scope.ClusterState, error)
+}
+
+// NewGenerator creates a new generator to generate desired state.
+func NewGenerator(client client.Client, tracker *remote.ClusterCacheTracker, runtimeClient runtimeclient.Client) Generator {
+	return &generator{
+		Client:        client,
+		Tracker:       tracker,
+		RuntimeClient: runtimeClient,
+		patchEngine:   patches.NewEngine(runtimeClient),
+	}
+}
+
+// generator is a generator to generate desired state.
+// It is used in the cluster topology controller, but it can also be used for testing.
+type generator struct {
+	Client client.Client
+
+	Tracker *remote.ClusterCacheTracker
+
+	RuntimeClient runtimeclient.Client
+
+	// patchEngine is used to apply patches during computeDesiredState.
+	patchEngine patches.Engine
+}
+
+// Generate computes the desired state of the cluster topology.
 // NOTE: We are assuming all the required objects are provided as input; also, in case of any error,
 // the entire compute operation will fail. This might be improved in the future if support for reconciling
 // subset of a topology will be implemented.
-func (r *Reconciler) computeDesiredState(ctx context.Context, s *scope.Scope) (*scope.ClusterState, error) {
+func (e *generator) Generate(ctx context.Context, s *scope.Scope) (*scope.ClusterState, error) {
```

Comment:
nit, optional, but I think g is more obvious to use.

```suggestion
func (g *generator) Generate(ctx context.Context, s *scope.Scope) (*scope.ClusterState, error) {
```

Reply:
Absolutely! Just forgot to rename after the rename of the struct (from engine before)